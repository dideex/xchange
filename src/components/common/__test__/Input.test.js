import React from 'react'
import Component from '../Input'
import {shallow, mount, render} from 'enzyme'
import WAValidator from 'coin-address-validator'

jest.mock('coin-address-validator', () => ({
  validate: () => {},
}))

describe('Input behaviour', () => {
  const testValue = 'Test value'
  const fakeMask = 'Btc'
  const testPhone = '89148906030'
  const testCardNumber = '1234123443214321'

  describe('Markups', () => {
      it('markup', () => {
        const wrapper = shallow(<Component handleChange={() => {}} />)
        expect(wrapper.html()).toMatchSnapshot()
      })

      it('markup value', () => {
        const wrapper = shallow(<Component handleChange={() => {}} value="My value" />)
        expect(wrapper.html()).toMatchSnapshot()
      })

      it('markup is invalid', () => {
        const wrapper = shallow(<Component handleChange={() => {}} isInvalid />)
        expect(wrapper.html()).toMatchSnapshot()
      })

      it('markup with error message', () => {
        const wrapper = shallow(<Component handleChange={() => {}} isInvalid errorMsg="Test message" />)
        expect(wrapper.html()).toMatchSnapshot()
      })

      it('markup with placeholder', () => {
        const wrapper = shallow(<Component handleChange={() => {}} placeholder="Test placeholder" />)
        expect(wrapper.html()).toMatchSnapshot()
      })

      it('markup with custom styles', () => {
        const wrapper = shallow(<Component handleChange={() => {}} style={{"testprop": "testKey"}} />)
        expect(wrapper.html()).toMatchSnapshot()
      })
  })

  describe('Values behaviour', () => {
    it('Value should be', () => {
      const wrapper = shallow(<Component handleChange={() => {}} value={testValue} />)
      expect(wrapper.html()).toContain(testValue)
    })

    it('Custom onBlur handler', () => {
      const onBlur = jest.fn()
      const wrapper = mount(<Component handleChange={() => {}} onBlur={onBlur} />)
      wrapper.find('input').simulate('blur')
      expect(onBlur).toHaveBeenCalledTimes(1)
    })

    it('Custom enter press event', () => {
      const handleEnterPress = jest.fn()
      const wrapper = mount(<Component handleChange={() => {}} handleEnterPress={handleEnterPress} />)
      wrapper.find('input').simulate('keypress', {which: 13})
      expect(handleEnterPress).toHaveBeenCalledTimes(1)
    })

    it('handleChange should be invoked on form change', async () => {
      const toggle = jest.fn()
      const wrapper = mount(<Component handleChange={toggle} value="" />)
      wrapper.find('input').simulate('change', {target: {value: testValue}})
      expect(toggle).toHaveBeenCalledTimes(1)
      expect(toggle).toHaveBeenCalledWith(testValue)
    })

    it('Get value from input first', async () => {
      const toggle = jest.fn()
      const wrapper = mount(<Component handleChange={toggle} value="Not used value" />)
      const instance = wrapper.instance()
      instance.handleChange({target: {value: testValue}})
      expect(toggle).toHaveBeenCalledTimes(1)
      expect(toggle).toHaveBeenCalledWith(testValue)
    })

    describe('Input validation', () => {
      it('Validate with mask success', () => {
        const toggle = jest.fn()
        const hanldeErrorChange = jest.fn()
        WAValidator.validate = jest.fn(() => true)
        const wrapper = mount(
          <Component
            handleErrorChange={hanldeErrorChange}
            mask={fakeMask}
            handleChange={toggle}
            value="Not used value"
          />,
        )
        const instance = wrapper.instance()
        instance.handleChange({target: {value: testValue}})
        expect(WAValidator.validate).toHaveBeenCalledTimes(1)
        expect(WAValidator.validate).toHaveBeenCalledWith(testValue, 'Btc')
        expect(hanldeErrorChange).toHaveBeenCalledTimes(1)
        expect(hanldeErrorChange).toHaveBeenCalledWith(false, expect.anything())
        expect(instance._validateWithMask(testValue, () => {})).toBe(testValue)
        expect(instance.mask).toMatchSnapshot()
      })

      it('Validate with mask unvalidation currencies', () => {
        const toggle = jest.fn()
        const hanldeErrorChange = jest.fn()
        WAValidator.validate = jest.fn(() => true)
        const wrapper = mount(
          <Component
            handleErrorChange={hanldeErrorChange}
            mask="LSK"
            handleChange={toggle}
            value="Not used value"
          />,
        )
        const instance = wrapper.instance()
        instance.handleChange({target: {value: testValue}})
        expect(WAValidator.validate).toHaveBeenCalledTimes(0)
        expect(hanldeErrorChange).toHaveBeenCalledTimes(0)
        expect(instance._validateWithMask(testValue, () => {})).toBe(testValue)
        expect(instance.mask).toMatchSnapshot()
      })

      it('Validate with mask phone', () => {
        const toggle = jest.fn()
        const hanldeErrorChange = jest.fn()
        WAValidator.validate = jest.fn(() => true)
        const wrapper = mount(
          <Component
            handleErrorChange={hanldeErrorChange}
            mask="phone"
            handleChange={toggle}
            value="Not used value"
          />,
        )
        const instance = wrapper.instance()
        instance.handleChange({target: {value: testValue}})
        expect(WAValidator.validate).toHaveBeenCalledTimes(0)
        expect(hanldeErrorChange).toHaveBeenCalledTimes(1)
        expect(hanldeErrorChange).toHaveBeenCalledWith(true, expect.anything())
        expect(instance._validateWithMask(testValue, () => {})).toMatchSnapshot()
        expect(instance.mask).toMatchSnapshot()
      })

      it('Validate card with pattern', () => {
        const toggle = jest.fn()
        const hanldeErrorChange = jest.fn()
        WAValidator.validate = jest.fn(() => true)
        const wrapper = mount(
          <Component
            handleErrorChange={hanldeErrorChange}
            mask="RUR"
            handleChange={toggle}
            value="Not used value"
            pattern="\d+"
          />,
        )
        const instance = wrapper.instance()
        instance.handleChange({target: {value: testCardNumber}})
        expect(WAValidator.validate).toHaveBeenCalledTimes(0)
        expect(hanldeErrorChange).toHaveBeenCalledTimes(1)
        expect(hanldeErrorChange).toHaveBeenCalledWith(false, expect.anything())
        expect(instance._validateWithMask(testCardNumber, () => {})).toMatchSnapshot()
        expect(instance.mask).toMatchSnapshot()
      })

      it('Validate card with pattern crash test', () => {
        const toggle = jest.fn()
        const hanldeErrorChange = jest.fn()
        const wrapper = mount(
          <Component
            handleErrorChange={hanldeErrorChange}
            mask="RUR"
            handleChange={toggle}
            value="Not used value"
            pattern="\d+"
          />,
        )
        const instance = wrapper.instance()
        expect(instance._validateWithMask('1234123412341234', () => {})).toMatchSnapshot()
        expect(instance._validateWithMask('12344321', () => {})).toMatchSnapshot()
        expect(instance._validateWithMask('12345678912345678123456', () => {})).toMatchSnapshot()
      })

      it('Validate phone with pattern', () => {
        const toggle = jest.fn()
        const hanldeErrorChange = jest.fn()
        WAValidator.validate = jest.fn(() => true)
        const wrapper = mount(
          <Component
            handleErrorChange={hanldeErrorChange}
            mask="phone"
            handleChange={toggle}
            value="Not used value"
            pattern="\d+"
          />,
        )
        const instance = wrapper.instance()
        instance.handleChange({target: {value: testPhone}})
        expect(WAValidator.validate).toHaveBeenCalledTimes(0)
        expect(hanldeErrorChange).toHaveBeenCalledTimes(1)
        expect(hanldeErrorChange).toHaveBeenCalledWith(false, expect.anything())
        expect(instance._validateWithMask(testPhone, () => {})).toMatchSnapshot()
        expect(instance.mask).toMatchSnapshot()
      })

      it('Validate phone with pattern crash test', () => {
        const toggle = jest.fn()
        const hanldeErrorChange = jest.fn()
        const wrapper = mount(
          <Component
            handleErrorChange={hanldeErrorChange}
            mask="phone"
            handleChange={toggle}
            value="Not used value"
            pattern="\d+"
          />,
        )
        const instance = wrapper.instance()
        expect(instance._validateWithMask('8914890300000', () => {})).toMatchSnapshot()
        expect(instance._validateWithMask('891489442', () => {})).toMatchSnapshot()
        expect(instance._validateWithMask('891489', () => {})).toMatchSnapshot()
      })
    })
    describe('Input formatt', () => {
      it('Phone format', () => {
        const wrapper = mount(
          <Component
            mask="phone"
            handleChange={() => {}}
            value=""
            pattern="\d+"
          />,
        )
        const instance = wrapper.instance()
        expect(instance._format('89148906034', () => {})).toMatchSnapshot()
        expect(instance._format('8914890603489148906034', () => {})).toMatchSnapshot()
        expect(instance._format('89148', () => {})).toMatchSnapshot()
        expect(instance._format('89148qwer', () => {})).toMatchSnapshot()
        expect(instance._format('89 1 48 906 0 34', () => {})).toMatchSnapshot()
      })

      it('Phone clean', () => {
        const wrapper = mount(
          <Component
            mask="phone"
            handleChange={() => {}}
            value=""
            pattern="\d+"
          />,
        )
        const instance = wrapper.instance()
        expect(instance._clean('8914_890_6034', () => {})).toMatchSnapshot()
        expect(instance._clean('891_48906034_891489_06034', () => {})).toMatchSnapshot()
        expect(instance._clean('8_9148', () => {})).toMatchSnapshot()
        expect(instance._clean('891_48_qwer', () => {})).toMatchSnapshot()
        expect(instance._clean('8_9 1 48_ 906 0 34', () => {})).toMatchSnapshot()
      })

      it('Card format', () => {
        const wrapper = mount(
          <Component
            mask="Btc"
            handleChange={() => {}}
            value=""
            pattern="\d+"
          />,
        )
        const instance = wrapper.instance()
        expect(instance._format('1234432164316423', () => {})).toMatchSnapshot()
        expect(instance._format('12344321643164231234432164316423', () => {})).toMatchSnapshot()
        expect(instance._format('12344321', () => {})).toMatchSnapshot()
        expect(instance._format('12344321qwe', () => {})).toMatchSnapshot()
        expect(instance._format('12 34 432 164 3164 23', () => {})).toMatchSnapshot()
      })

      it('Card clean', () => {
        const wrapper = mount(
          <Component
            mask="Btc"
            handleChange={() => {}}
            value=""
            pattern="\d+"
          />,
        )
        const instance = wrapper.instance()
        expect(instance._format('1_2344_32_1643_164_23', () => {})).toMatchSnapshot()
        expect(instance._format('1234_432__16431_6423123_443216_4316423', () => {})).toMatchSnapshot()
        expect(instance._format('12_3443_21', () => {})).toMatchSnapshot()
        expect(instance._format('123_4432_1qw_e', () => {})).toMatchSnapshot()
        expect(instance._format('1_2 34 43__ 1_64 3_164 _23', () => {})).toMatchSnapshot()
      })
    })
  })
})
