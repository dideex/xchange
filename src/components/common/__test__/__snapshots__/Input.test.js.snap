// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input behaviour Markups markup 1`] = `"<label class=\\"css-gudk1o\\"><input type=\\"text\\" value=\\"\\" placeholder=\\"\\" mask=\\"\\" class=\\"css-18b8ap6\\"/></label>"`;

exports[`Input behaviour Markups markup is invalid 1`] = `"<label class=\\"css-gudk1o\\"><input type=\\"text\\" value=\\"\\" placeholder=\\"\\" mask=\\"\\" class=\\"css-ykrnj4\\"/><span class=\\"css-1j8ju8e\\"></span></label>"`;

exports[`Input behaviour Markups markup value 1`] = `"<label class=\\"css-gudk1o\\"><input type=\\"text\\" value=\\"My value\\" placeholder=\\"\\" mask=\\"\\" class=\\"css-18b8ap6\\"/></label>"`;

exports[`Input behaviour Markups markup with custom styles 1`] = `"<label class=\\"css-gudk1o\\"><input type=\\"text\\" style=\\"testprop:testKey\\" value=\\"\\" placeholder=\\"\\" mask=\\"\\" class=\\"css-18b8ap6\\"/></label>"`;

exports[`Input behaviour Markups markup with error message 1`] = `"<label class=\\"css-gudk1o\\"><input type=\\"text\\" value=\\"\\" placeholder=\\"\\" mask=\\"\\" class=\\"css-ykrnj4\\"/><span class=\\"css-1j8ju8e\\">Test message</span></label>"`;

exports[`Input behaviour Markups markup with placeholder 1`] = `"<label class=\\"css-gudk1o\\"><input type=\\"text\\" placeholder=\\"Test placeholder\\" value=\\"\\" mask=\\"\\" class=\\"css-18b8ap6\\"/></label>"`;

exports[`Input behaviour Values behaviour Input formatt Card clean 1`] = `"1234 4321 6431 ____"`;

exports[`Input behaviour Values behaviour Input formatt Card clean 2`] = `"1234 4321 6431 ____"`;

exports[`Input behaviour Values behaviour Input formatt Card clean 3`] = `"1234 4321"`;

exports[`Input behaviour Values behaviour Input formatt Card clean 4`] = `"1234 4321 qwe"`;

exports[`Input behaviour Values behaviour Input formatt Card clean 5`] = `"12 3 4 43  164 ____"`;

exports[`Input behaviour Values behaviour Input formatt Card format 1`] = `"1234 4321 6431 6423"`;

exports[`Input behaviour Values behaviour Input formatt Card format 2`] = `"1234 4321 6431 6423"`;

exports[`Input behaviour Values behaviour Input formatt Card format 3`] = `"1234 4321"`;

exports[`Input behaviour Values behaviour Input formatt Card format 4`] = `"1234 4321 qwe"`;

exports[`Input behaviour Values behaviour Input formatt Card format 5`] = `"12 3 4 43 2 16 4 31"`;

exports[`Input behaviour Values behaviour Input formatt Phone clean 1`] = `"89148906034"`;

exports[`Input behaviour Values behaviour Input formatt Phone clean 2`] = `"8914890603489148906034"`;

exports[`Input behaviour Values behaviour Input formatt Phone clean 3`] = `"89148"`;

exports[`Input behaviour Values behaviour Input formatt Phone clean 4`] = `"89148qwer"`;

exports[`Input behaviour Values behaviour Input formatt Phone clean 5`] = `"89 1 48 906 0 34"`;

exports[`Input behaviour Values behaviour Input formatt Phone format 1`] = `"+8(914)-890-60-34"`;

exports[`Input behaviour Values behaviour Input formatt Phone format 2`] = `"+8(914)-890-60-34"`;

exports[`Input behaviour Values behaviour Input formatt Phone format 3`] = `"+8(914)-8"`;

exports[`Input behaviour Values behaviour Input formatt Phone format 4`] = `"+8(914)-8qw-er"`;

exports[`Input behaviour Values behaviour Input formatt Phone format 5`] = `"+8(9 1)- 48- 9-06"`;

exports[`Input behaviour Values behaviour Input validation Validate card with pattern 1`] = `"1234 1234 4321 4321"`;

exports[`Input behaviour Values behaviour Input validation Validate card with pattern 2`] = `"____ ____ ____ ____"`;

exports[`Input behaviour Values behaviour Input validation Validate card with pattern crash test 1`] = `"1234 1234 1234 1234"`;

exports[`Input behaviour Values behaviour Input validation Validate card with pattern crash test 2`] = `"1234 4321"`;

exports[`Input behaviour Values behaviour Input validation Validate card with pattern crash test 3`] = `"1234 5678 9123 4567"`;

exports[`Input behaviour Values behaviour Input validation Validate phone with pattern 1`] = `"+8(914)-890-60-30"`;

exports[`Input behaviour Values behaviour Input validation Validate phone with pattern 2`] = `"+_(___)-___-__-__"`;

exports[`Input behaviour Values behaviour Input validation Validate phone with pattern crash test 1`] = `"+8(914)-890-30-00"`;

exports[`Input behaviour Values behaviour Input validation Validate phone with pattern crash test 2`] = `"+8(914)-894-42"`;

exports[`Input behaviour Values behaviour Input validation Validate phone with pattern crash test 3`] = `"+8(914)-89"`;

exports[`Input behaviour Values behaviour Input validation Validate with mask phone 1`] = `"+T(est)- va-lu-e"`;

exports[`Input behaviour Values behaviour Input validation Validate with mask phone 2`] = `"+_(___)-___-__-__"`;

exports[`Input behaviour Values behaviour Input validation Validate with mask success 1`] = `"____ ____ ____ ____"`;

exports[`Input behaviour Values behaviour Input validation Validate with mask unvalidation currencies 1`] = `"____ ____ ____ ____"`;
